### ISSUE TITLE
Linux GTK - OSMesa doesn't work

### ISSUE THREAD

**Issue by OrdinaryMagician** [TASVideos/desmume#119]
I was wondering why the renderer always flipped back to software whenever I tried to change it. Looking at the source I see there isn't even any GL support in the first place, just a bare skeleton that immediately fails because it's not even creating a context or anything.

Has no one else noticed this before?

**Comment by rogerman** [comment]
Works for me.

The GTK port has the option to use either OSMesa or GLX contexts. Ensure that you are compiling with the correct OpenGL libraries and try again.

Closing, due to user error.

**Comment by OrdinaryMagician** [comment]
I built with --enable-osmesa and --enable-glx. I have the required libraries. I set the renderer to OpenGL. It reverted to software. I got the message `OSMesaCreateContext failed!` on stdout. I don't see how this is user error.

**Comment by OrdinaryMagician** [comment]
I've looked at the code again and I see that OSMesa takes priority and GLX is never even used if both are compiled in. The thing is, as I said initially, the OSMesa renderer doesn't even create a context or anything and just fails immediately [as seen here](https://github.com/TASVideos/desmume/blob/b9ada994dfb6c4143fe077d61394dca5c7e6fd8f/desmume/src/frontend/posix/gtk/osmesa_3Demu.cpp#L42-L46).

So, let me rephrase this whole issue: OSMesa doesn't work.

PS: I'm sorry if I sound angry.

**Comment by zeromus** [comment]
You shouldn't use both those options.
Try changing gtk's main.cpp line 2264 to `#endif` && `#if defined(HAVE_GL_GLX)` and the whole block at line 3371 to
```
#if defined(HAVE_LIBOSMESA) || defined(HAVE_GL_GLX)
    if(core == 2)
    {
#if defined(HAVE_LIBOSMESA)
        core = init_osmesa_3Demu() ? 2 : RENDERID_NULL;
#endif
    }
    if(core == 2) 
  {
#if defined(HAVE_GL_GLX)
        core = init_glx_3Demu() ? 2 : RENDERID_NULL;
#endif
    }
#endif
```
and line 3448 to `#endif` && `#if defined(HAVE_GL_GLX)`

It looks possible to me that with this we could have one work if the other fails, but I won't debug it any further than this

**Comment by OrdinaryMagician** [comment]
I've tested with those changes but it seems to still fall back to software.

For now I'll build just with GLX, since that one works perfectly fine.

**Comment by zeromus** [comment]
Those changes won't work til @klapeto 's problem is resolved

**Comment by zeromus** [comment]
my changes weren't right. let me try again. 
```
    if(core == 2)
    {
bool ok = false;
#if defined(HAVE_LIBOSMESA)
ok = !!init_osmesa_3Demu();
#endif
#if defined(HAVE_GL_GLX)
if(!ok)
        ok = !!init_osmesa_3Demu();
#endif
core = ok ? 2 : RENDERID_NULL;
    }

**Comment by OrdinaryMagician** [comment]
Didn't work either. After some debugging I've noticed that HAVE_GL_GLX is not defined even though I also passed --enable-glx on configure. Seems that there's something that forcibly disables it if osmesa is enabled.

**Comment by zeromus** [comment]
ok, so we have another issue if the flags go through configure OK but then result in dysfunctional builds

**Comment by rogerman** [comment]
You're misunderstanding something here -- GLX and OSMesa are mutually exclusive. It is intentional for --enable-osmesa to override GLX. If you use --enable-glx with --enable-osmesa, then --enable-glx will have no effect.

By default, GLX will be used if the GLX libraries exist. There is no need to add --enable-glx because it is automatically enabled by default (unless overridden by --enable-osmesa).

In any case, OSMesa should be fixed now as of commit 010efff. The problem was a longstanding regression where the OSMesa context wasn't even being created in the first place.
